File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int len = nums.size();
        if (len == 1) {
            return nums[0];
        } else if (len == 2) {
            return max(nums[0], nums[1]);
        }

        return max(rob(nums, 0, len - 2), rob(nums, 1, len - 1));
    }

    int rob(vector<int>& nums, int start, int end) {
        if (start == end) {
            return nums[start];
        }

        int x = nums[start];
        int y = max(x, nums[start + 1]);
        int z = 0;
        for (int i = start + 2; i <= end; i++) {
            z = max(y, x + nums[i]);
            x = y;
            y = z;

        }

        return y;
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]

        def helper(start, end) -> int:
            if start == end:
                return nums[start]
            x = nums[start]
            y = max(nums[start], nums[start + 1])

            for i in range(start + 2, end + 1):
                z = y
                y = max(x + nums[i], y)
                x = z

            return max(x, y)


        return max(helper(0, n - 2), helper(1, n - 1))